src/System.Net.Security/src/System/Net/_SecureChannel.cs

```diff
@@ -0,0 +1,1487 @@
+
+            ChannelBinding result = null;
+            if (_securityContext != null)
+            {
+                result = SSPIWrapper.QueryContextChannelBinding(GlobalSSPI.SSPISecureChannel, _securityContext, (Interop.Secur32.ContextAttribute)kind);
+            }
+
+            GlobalLog.Leave("SecureChannel#" + Logging.HashString(this) + "::GetChannelBindingToken", Logging.HashString(result));
+            return result;
+        }
+
+        internal bool CheckCertRevocationStatus
+        {
+            get
+            {
+                return _checkCertRevocation;
+            }
+        }
+
+        internal X509CertificateCollection ClientCertificates
+        {
+            get
+            {
+                return _clientCertificates;
+            }
+        }
+
+        internal int HeaderSize
+        {
+            get
+            {
+                return _headerSize;
+            }
+        }
+
+        internal int MaxDataSize
+        {
+            get
+            {
+                return _maxDataSize;
+            }
+        }
+
+        internal SslConnectionInfo ConnectionInfo
+        {
+            get
+            {
+                return _connectionInfo;
+            }
+        }
+
+        internal bool IsValidContext
+        {
+            get
+            {
+                return !(_securityContext == null || _securityContext.IsInvalid);
+            }
+        }
+
+        internal bool IsServer
+        {
+            get
+            {
+                return _serverMode;
+            }
+        }
+
+        internal bool RemoteCertRequired
+        {
+            get
+            {
+                return _remoteCertRequired;
+            }
+        }
+
+        internal void SetRefreshCredentialNeeded()
+        {
+            _refreshCredentialNeeded = true;
+        }
+
+        internal void Close()
+        {
+            if (_securityContext != null)
+            {
+                _securityContext.Dispose();
+            }
+            if (_credentialsHandle != null)
+            {
+                _credentialsHandle.Dispose();
+            }
+        }
+
+        //
+        // SECURITY: we open a private key container on behalf of the caller
+        // and we require the caller to have permission associated with that operation.
+        //
+        private X509Certificate2 EnsurePrivateKey(X509Certificate certificate)
+        {
+            if (certificate == null)
+                return null;
+
+            if (Logging.On) Logging.PrintInfo(Logging.Web, this, SR.Format(SR.net_log_locating_private_key_for_certificate, certificate.ToString(true)));
+
+            try
+            {
+                X509Certificate2 certEx = certificate as X509Certificate2;
+                Type t = certificate.GetType();
+                string certHash = null;
+
+                // Protecting from X509Certificate2 derived classes.
+                if (t != typeof(X509Certificate2))
```

Comment:
This looks duplicate to the `MakeEx` method below, could the first half of the try not just be replaced with
```C#
X509Certificate2 certEx = MakeEx(certificate);
// A nullcheck might be prudent, since MakeEx eats the exceptions and returns null, and this would eat it differently.
string certHash = certEx.Thumbprint;
```